File: ios/Sources/SocialLoginPlugin/TwitterProvider.swift
import Foundation
import AuthenticationServices
import CryptoKit
import UIKit

struct TwitterTokenResponse: Codable {
    let token_type: String
    let expires_in: Int
    let access_token: String
    let scope: String
    let refresh_token: String?
}

struct TwitterUserEnvelope: Codable {
    struct TwitterUserData: Codable {
        let id: String
        let name: String?
        let username: String
        let profile_image_url: String?
        let verified: Bool?
        let email: String?
    }

    let data: TwitterUserData
}

struct TwitterProfileResponse {
    let accessToken: TwitterAccessToken
    let profile: TwitterUserEnvelope.TwitterUserData
    let scope: [String]
    let expiresIn: Int
    let tokenType: String
}

struct TwitterAccessToken: Codable {
    let token: String
    let expiresIn: Int?
    let refreshToken: String?
    let tokenType: String?
}

struct TwitterStoredTokens: Codable {
    let accessToken: String
    let refreshToken: String?
    let expiresAt: Date
    let scope: [String]
    let tokenType: String
    let userId: String?
}

class TwitterProvider: NSObject {
    private var clientId: String?
    private var redirectUri: String?
    private var defaultScopes = ["tweet.read", "users.read"]
    private var forceLogin = false
    private var audience: String?

    private var currentSession: ASWebAuthenticationSession?
    private var currentState: String?
    private let tokenStorageKey = "CapgoTwitterProviderTokens"

    func initialize(clientId: String, redirectUri: String, defaultScopes: [String]?, forceLogin: Bool, audience: String?) {
        self.clientId = clientId
        self.redirectUri = redirectUri
        if let scopes = defaultScopes, !scopes.isEmpty {
            self.defaultScopes = scopes
        }
        self.forceLogin = forceLogin
        self.audience = audience
    }

    func login(payload: [String: Any], completion: @escaping (Result<TwitterProfileResponse, Error>) -> Void) {
        guard let clientId = clientId, let redirectUri = redirectUri else {
            completion(.failure(NSError(domain: "TwitterProvider", code: -1, userInfo: [NSLocalizedDescriptionKey: "Twitter client not configured. Call initialize()."])))
            return
        }

        let scopes = payload["scopes"] as? [String] ?? defaultScopes
        let state = payload["state"] as? String ?? UUID().uuidString
        let codeVerifier = payload["codeVerifier"] as? String ?? generateCodeVerifier()
        let forceLoginFlag = payload["forceLogin"] as? Bool ?? forceLogin
        let redirect = payload["redirectUrl"] as? String ?? redirectUri

        currentState = state

        var components = URLComponents(string: "https://x.com/i/oauth2/authorize")
        var query: [URLQueryItem] = [
            URLQueryItem(name: "response_type", value: "code"),
            URLQueryItem(name: "client_id", value: clientId),
            URLQueryItem(name: "redirect_uri", value: redirect),
            URLQueryItem(name: "scope", value: scopes.joined(separator: " ")),
            URLQueryItem(name: "state", value: state),
            URLQueryItem(name: "code_challenge", value: generateCodeChallenge(from: codeVerifier)),
            URLQueryItem(name: "code_challenge_method", value: "S256")
        ]
        if forceLoginFlag {
            query.append(URLQueryItem(name: "force_login", value: "true"))
        }
        if let audience = audience {
            query.append(URLQueryItem(name: "audience", value: audience))
        }
        components?.queryItems = query

        guard let authUrl = components?.url, let callbackScheme = URL(string: redirect)?.scheme else {
            completion(.failure(NSError(domain: "TwitterProvider", code: -2, userInfo: [NSLocalizedDescriptionKey: "Invalid redirect URL configuration."])) )
            return
        }

        let session = ASWebAuthenticationSession(url: authUrl, callbackURLScheme: callbackScheme) { [weak self] callbackURL, error in
            guard let self = self else { return }
            if let error = error {
                completion(.failure(error))
                return
            }
            guard let callbackURL = callbackURL, let queryItems = URLComponents(url: callbackURL, resolvingAgainstBaseURL: false)?.queryItems else {
                completion(.failure(NSError(domain: "TwitterProvider", code: -3, userInfo: [NSLocalizedDescriptionKey: "Invalid callback URL."])) )
                return
            }

            let returnedState = queryItems.first(where: { $0.name == "state" })?.value
            if returnedState != self.currentState {
                completion(.failure(NSError(domain: "TwitterProvider", code: -4, userInfo: [NSLocalizedDescriptionKey: "State mismatch detected during Twitter login."])) )
                return
            }

            if let errorCode = queryItems.first(where: { $0.name == "error" })?.value {
                let errorDescription = queryItems.first(where: { $0.name == "error_description" })?.value ?? errorCode
                completion(.failure(NSError(domain: "TwitterProvider", code: -5, userInfo: [NSLocalizedDescriptionKey: errorDescription])))
                return
            }

            guard let code = queryItems.first(where: { $0.name == "code" })?.value else {
                completion(.failure(NSError(domain: "TwitterProvider", code: -6, userInfo: [NSLocalizedDescriptionKey: "Authorization code missing."])) )
                return
            }

            self.exchangeCode(code: code, redirectUri: redirect, codeVerifier: codeVerifier) { exchangeResult in
                completion(exchangeResult)
            }
        }

        if #available(iOS 13.0, *) {
            session.presentationContextProvider = self
        }

        session.prefersEphemeralWebBrowserSession = true
        currentSession = session
        session.start()
    }

    func logout(completion: @escaping (Result<Void, Error>) -> Void) {
        UserDefaults.standard.removeObject(forKey: tokenStorageKey)
        completion(.success(()))
    }

    func isLoggedIn(completion: @escaping (Result<Bool, Error>) -> Void) {
        if let tokens = loadTokens() {
            completion(.success(tokens.expiresAt > Date()))
        } else {
            completion(.success(false))
        }
    }

    func getAuthorizationCode(completion: @escaping (Result<TwitterAccessToken, Error>) -> Void) {
        guard let tokens = loadTokens() else {
            completion(.failure(NSError(domain: "TwitterProvider", code: -7, userInfo: [NSLocalizedDescriptionKey: "User not logged in"])))
            return
        }
        completion(.success(TwitterAccessToken(token: tokens.accessToken, expiresIn: Int(tokens.expiresAt.timeIntervalSince(Date())), refreshToken: tokens.refreshToken, tokenType: tokens.tokenType)))
    }

    func refresh(completion: @escaping (Result<TwitterProfileResponse, Error>) -> Void) {
        guard let tokens = loadTokens(), let refreshToken = tokens.refreshToken else {
            completion(.failure(NSError(domain: "TwitterProvider", code: -8, userInfo: [NSLocalizedDescriptionKey: "No refresh token available. Include offline.access scope."])) )
            return
        }

        refreshTokens(refreshToken: refreshToken) { result in
            completion(result)
        }
    }

    private func exchangeCode(code: String, redirectUri: String, codeVerifier: String, completion: @escaping (Result<TwitterProfileResponse, Error>) -> Void) {
        guard let clientId = clientId else {
            completion(.failure(NSError(domain: "TwitterProvider", code: -9, userInfo: [NSLocalizedDescriptionKey: "Missing client id"])))
            return
        }

        let body = [
            "grant_type": "authorization_code",
            "client_id": clientId,
            "code": code,
            "redirect_uri": redirectUri,
            "code_verifier": codeVerifier
        ]

        performTokenRequest(body: body) { [weak self] result in
            switch result {
            case .success(let tokenResponse):
                self?.handleTokenSuccess(tokenResponse: tokenResponse, completion: completion)
            case .failure(let error):
                completion(.failure(error))
            }
        }
    }

    private func refreshTokens(refreshToken: String, completion: @escaping (Result<TwitterProfileResponse, Error>) -> Void) {
        guard let clientId = clientId else {
            completion(.failure(NSError(domain: "TwitterProvider", code: -9, userInfo: [NSLocalizedDescriptionKey: "Missing client id"])))
            return
        }

        let body = [
            "grant_type": "refresh_token",
            "refresh_token": refreshToken,
            "client_id": clientId
        ]

        performTokenRequest(body: body) { [weak self] result in
            switch result {
            case .success(let tokenResponse):
                self?.handleTokenSuccess(tokenResponse: tokenResponse, completion: completion)
            case .failure(let error):
                completion(.failure(error))
            }
        }
    }

    private func performTokenRequest(body: [String: String], completion: @escaping (Result<TwitterTokenResponse, Error>) -> Void) {
        guard let url = URL(string: "https://api.x.com/2/oauth2/token") else {
            completion(.failure(NSError(domain: "TwitterProvider", code: -10, userInfo: [NSLocalizedDescriptionKey: "Invalid token endpoint."])) )
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.addValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.httpBody = body.percentEncoded()

        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            guard let httpResponse = response as? HTTPURLResponse else {
                completion(.failure(NSError(domain: "TwitterProvider", code: -11, userInfo: [NSLocalizedDescriptionKey: "Invalid response"])))
                return
            }
            guard let data = data, httpResponse.statusCode == 200 else {
                let message = String(data: data ?? Data(), encoding: .utf8) ?? "Unknown error"
                completion(.failure(NSError(domain: "TwitterProvider", code: httpResponse.statusCode, userInfo: [NSLocalizedDescriptionKey: message])))
                return
            }

            do {
                let decoded = try JSONDecoder().decode(TwitterTokenResponse.self, from: data)
                completion(.success(decoded))
            } catch {
                completion(.failure(error))
            }
        }
        .resume()
    }

    private func handleTokenSuccess(tokenResponse: TwitterTokenResponse, completion: @escaping (Result<TwitterProfileResponse, Error>) -> Void) {
        fetchProfile(accessToken: tokenResponse.access_token) { [weak self] profileResult in
            switch profileResult {
            case .success(let profile):
                let expiresAt = Date().addingTimeInterval(TimeInterval(tokenResponse.expires_in))
                let scopeArray = tokenResponse.scope.split(separator: " ").map { String($0) }

                let stored = TwitterStoredTokens(
                    accessToken: tokenResponse.access_token,
                    refreshToken: tokenResponse.refresh_token,
                    expiresAt: expiresAt,
                    scope: scopeArray,
                    tokenType: tokenResponse.token_type,
                    userId: profile.id
                )
                self?.persistTokens(tokens: stored)

                let response = TwitterProfileResponse(
                    accessToken: TwitterAccessToken(token: tokenResponse.access_token, expiresIn: tokenResponse.expires_in, refreshToken: tokenResponse.refresh_token, tokenType: tokenResponse.token_type),
                    profile: profile,
                    scope: scopeArray,
                    expiresIn: tokenResponse.expires_in,
                    tokenType: tokenResponse.token_type
                )
                completion(.success(response))
            case .failure(let error):
                completion(.failure(error))
            }
        }
    }

    private func fetchProfile(accessToken: String, completion: @escaping (Result<TwitterUserEnvelope.TwitterUserData, Error>) -> Void) {
        var components = URLComponents(string: "https://api.x.com/2/users/me")
        components?.queryItems = [URLQueryItem(name: "user.fields", value: "profile_image_url,verified,name,username")]

        guard let url = components?.url else {
            completion(.failure(NSError(domain: "TwitterProvider", code: -12, userInfo: [NSLocalizedDescriptionKey: "Invalid user endpoint."])) )
            return
        }

        var request = URLRequest(url: url)
        request.addValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")

        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            guard let httpResponse = response as? HTTPURLResponse else {
                completion(.failure(NSError(domain: "TwitterProvider", code: -13, userInfo: [NSLocalizedDescriptionKey: "Invalid response"])))
                return
            }
            guard let data = data, httpResponse.statusCode == 200 else {
                let message = String(data: data ?? Data(), encoding: .utf8) ?? "Unknown error"
                completion(.failure(NSError(domain: "TwitterProvider", code: httpResponse.statusCode, userInfo: [NSLocalizedDescriptionKey: message])))
                return
            }
            do {
                let decoded = try JSONDecoder().decode(TwitterUserEnvelope.self, from: data)
                completion(.success(decoded.data))
            } catch {
                completion(.failure(error))
            }
        }
        .resume()
    }

    private func persistTokens(tokens: TwitterStoredTokens) {
        do {
            let data = try JSONEncoder().encode(tokens)
            UserDefaults.standard.set(data, forKey: tokenStorageKey)
        } catch {
            print("TwitterProvider persistTokens error: \(error.localizedDescription)")
        }
    }

    private func loadTokens() -> TwitterStoredTokens? {
        guard let data = UserDefaults.standard.data(forKey: tokenStorageKey) else {
            return nil
        }
        do {
            return try JSONDecoder().decode(TwitterStoredTokens.self, from: data)
        } catch {
            print("TwitterProvider loadTokens error: \(error.localizedDescription)")
            return nil
        }
    }

    private func generateCodeVerifier() -> String {
        var data = Data(count: 64)
        _ = data.withUnsafeMutableBytes { SecRandomCopyBytes(kSecRandomDefault, 64, $0.baseAddress!) }
        return data.base64EncodedString().replacingOccurrences(of: "=", with: "").replacingOccurrences(of: "+", with: "-").replacingOccurrences(of: "/", with: "_")
    }

    private func generateCodeChallenge(from verifier: String) -> String {
        let data = verifier.data(using: .utf8) ?? Data()
        let digest = SHA256.hash(data: data)
        return Data(digest).base64EncodedString().replacingOccurrences(of: "+", with: "-").replacingOccurrences(of: "/", with: "_").replacingOccurrences(of: "=", with: "")
    }
}

extension TwitterProvider: ASWebAuthenticationPresentationContextProviding {
    @available(iOS 13.0, *)
    func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
        return UIApplication.shared.windows.first { $0.isKeyWindow } ?? ASPresentationAnchor()
    }
}

private extension Dictionary where Key == String, Value == String {
    func percentEncoded() -> Data? {
        map { key, value in
            "\(key.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? key)=\(value.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? value)"
        }
        .joined(separator: "&")
        .data(using: .utf8)
    }
}

File: src/twitter-provider.ts
import { BaseSocialLogin } from './base';
import type {
  AuthorizationCode,
  LoginResult,
  ProviderResponseMap,
  TwitterLoginOptions,
  TwitterLoginResponse,
  TwitterProfile,
} from './definitions';

interface TwitterTokenResponse {
  token_type: 'bearer';
  expires_in: number;
  access_token: string;
  scope: string;
  refresh_token?: string;
}

interface TwitterUserResponse {
  data: {
    id: string;
    name: string;
    username: string;
    profile_image_url?: string;
    verified?: boolean;
  } & { email?: string };
}

interface TwitterPendingLogin {
  codeVerifier: string;
  redirectUri: string;
  scopes: string[];
}

interface TwitterStoredTokens {
  accessToken: string;
  refreshToken?: string;
  expiresAt: number;
  scope: string[];
  tokenType: string;
  userId?: string;
  profile?: TwitterProfile;
}

export class TwitterSocialLogin extends BaseSocialLogin {
  private clientId: string | null = null;
  private redirectUrl: string | null = null;
  private defaultScopes: string[] = ['tweet.read', 'users.read'];
  private forceLogin = false;
  private audience?: string;
  private readonly TOKENS_KEY = 'capgo_social_login_twitter_tokens_v1';
  private readonly STATE_PREFIX = 'capgo_social_login_twitter_state_';

  async initialize(
    clientId: string | null,
    redirectUrl?: string | null,
    defaultScopes?: string[],
    forceLogin?: boolean,
    audience?: string,
  ): Promise<void> {
    this.clientId = clientId;
    this.redirectUrl = redirectUrl ?? null;
    if (defaultScopes?.length) {
      this.defaultScopes = defaultScopes;
    }
    this.forceLogin = forceLogin ?? false;
    this.audience = audience ?? undefined;
  }

  async login<T extends 'twitter'>(
    options: TwitterLoginOptions,
  ): Promise<{ provider: T; result: ProviderResponseMap[T] }> {
    if (!this.clientId) {
      throw new Error('Twitter Client ID not configured. Call initialize() first.');
    }

    const redirectUri = options.redirectUrl ?? this.redirectUrl ?? window.location.origin + window.location.pathname;
    const scopes = options.scopes?.length ? options.scopes : this.defaultScopes;
    const state = options.state ?? this.generateState();
    const codeVerifier = options.codeVerifier ?? this.generateCodeVerifier();
    const codeChallenge = await this.generateCodeChallenge(codeVerifier);

    this.persistPendingLogin(state, {
      codeVerifier,
      redirectUri,
      scopes,
    });

    localStorage.setItem(BaseSocialLogin.OAUTH_STATE_KEY, JSON.stringify({ provider: 'twitter', state }));

    const params = new URLSearchParams({
      response_type: 'code',
      client_id: this.clientId,
      redirect_uri: redirectUri,
      scope: scopes.join(' '),
      state,
      code_challenge: codeChallenge,
      code_challenge_method: 'S256',
    });
    if ((options.forceLogin ?? this.forceLogin) === true) {
      params.set('force_login', 'true');
    }
    if (this.audience) {
      params.set('audience', this.audience);
    }

    const authUrl = `https://x.com/i/oauth2/authorize?${params.toString()}`;
    const width = 500;
    const height = 650;
    const left = window.screenX + (window.outerWidth - width) / 2;
    const top = window.screenY + (window.outerHeight - height) / 2;

    const popup = window.open(authUrl, 'XLogin', `width=${width},height=${height},left=${left},top=${top},popup=1`);

    return new Promise((resolve, reject) => {
      if (!popup) {
        reject(new Error('Unable to open login window. Please allow popups.'));
        return;
      }

      // Use BroadcastChannel for cross-origin communication (works when postMessage doesn't)
      const channelName = `twitter_oauth_${state}`;
      let broadcastChannel: BroadcastChannel | null = null;

      try {
        broadcastChannel = new BroadcastChannel(channelName);
      } catch {
        // BroadcastChannel not supported, fall back to postMessage only
      }

      // Track if we've encountered a COOP error to avoid repeated checks
      let coopErrorDetected = false;

      const cleanup = (
        messageHandler: (event: MessageEvent) => void,
        timeoutHandle: number,
        intervalHandle: number,
      ) => {
        window.removeEventListener('message', messageHandler);
        clearTimeout(timeoutHandle);
        clearInterval(intervalHandle);
        if (broadcastChannel) {
          broadcastChannel.close();
        }
      };

      const handleOAuthMessage = (data: Record<string, unknown>) => {
        if (data?.type === 'oauth-response') {
          if (data?.provider && data.provider !== 'twitter') {
            return false;
          }
          cleanup(messageHandler, timeoutHandle, popupClosedInterval);
          // eslint-disable-next-line @typescript-eslint/no-unused-vars
          const {
            provider: _ignoredProvider,
            type: _ignoredType,
            ...payload
          } = data as unknown as TwitterLoginResponse & {
            provider?: string;
            type?: string;
          };
          resolve({
            provider: 'twitter' as T,
            result: payload as ProviderResponseMap[T],
          } as { provider: T; result: ProviderResponseMap[T] });
          return true;
        } else if (data?.type === 'oauth-error') {
          if (data?.provider && data.provider !== 'twitter') {
            return false;
          }
          cleanup(messageHandler, timeoutHandle, popupClosedInterval);
          reject(new Error((data.error as string) || 'Twitter login was cancelled.'));
          return true;
        }
        return false;
      };

      // Listen for BroadcastChannel messages
      if (broadcastChannel) {
        broadcastChannel.onmessage = (event: MessageEvent) => {
          handleOAuthMessage(event.data);
        };
      }

      const messageHandler = (event: MessageEvent) => {
        if (event.origin !== window.location.origin) {
          return;
        }
        handleOAuthMessage(event.data);
      };

      window.addEventListener('message', messageHandler);

      const timeoutHandle = window.setTimeout(() => {
        cleanup(messageHandler, timeoutHandle, popupClosedInterval);
        try {
          popup.close();
        } catch {
          // Ignore cross-origin errors when closing
        }
        reject(new Error('Twitter login timed out.'));
      }, 300000);

      const popupClosedInterval = window.setInterval(() => {
        // Skip checking if we've already detected a COOP error
        if (coopErrorDetected) {
          return;
        }

        try {
          // Check if popup is closed - this may throw cross-origin errors for some providers
          if (popup.closed) {
            cleanup(messageHandler, timeoutHandle, popupClosedInterval);
            reject(new Error('Twitter login window was closed.'));
          }
        } catch {
          // Cross-origin error when checking popup.closed - this happens when the popup
          // navigates to a third-party OAuth provider with strict security settings (COOP).
          // We can't detect if the window was closed, so we rely on timeout and message handlers.
          // The popup will close itself after authentication completes.
          coopErrorDetected = true;
          console.log(
            '[Twitter Login] Cannot check popup.closed due to Cross-Origin-Opener-Policy restrictions. ' +
              'The popup will close automatically after login completes. Relying on BroadcastChannel and timeout.',
          );
        }
      }, 1000);
    });
  }

  async logout(): Promise<void> {
    localStorage.removeItem(this.TOKENS_KEY);
  }

  async isLoggedIn(): Promise<{ isLoggedIn: boolean }> {
    const tokens = this.getStoredTokens();
    if (!tokens) {
      return { isLoggedIn: false };
    }
    const isValid = tokens.expiresAt > Date.now();
    if (!isValid) {
      localStorage.removeItem(this.TOKENS_KEY);
    }
    return { isLoggedIn: isValid };
  }

  async getAuthorizationCode(): Promise<AuthorizationCode> {
    const tokens = this.getStoredTokens();
    if (!tokens) {
      throw new Error('Twitter access token is not available.');
    }
    return {
      accessToken: tokens.accessToken,
    };
  }

  async refresh(): Promise<void> {
    const tokens = this.getStoredTokens();
    if (!tokens?.refreshToken) {
      throw new Error('No Twitter refresh token is available. Include offline.access scope to receive one.');
    }

    await this.refreshWithRefreshToken(tokens.refreshToken);
  }

  async handleOAuthRedirect(url: URL, expectedState?: string): Promise<LoginResult | { error: string } | null> {
    const params = url.searchParams;
    const stateFromUrl = expectedState ?? params.get('state');
    if (!stateFromUrl) {
      return null;
    }

    const pending = this.consumePendingLogin(stateFromUrl);
    if (!pending) {
      localStorage.removeItem(BaseSocialLogin.OAUTH_STATE_KEY);
      return { error: 'Twitter login session expired or state mismatch.' };
    }

    const error = params.get('error');
    if (error) {
      localStorage.removeItem(BaseSocialLogin.OAUTH_STATE_KEY);
      return { error: params.get('error_description') || error };
    }

    const code = params.get('code');
    if (!code) {
      localStorage.removeItem(BaseSocialLogin.OAUTH_STATE_KEY);
      return { error: 'Twitter authorization code missing from redirect.' };
    }

    try {
      const tokens = await this.exchangeAuthorizationCode(code, pending);
      const profile = await this.fetchProfile(tokens.access_token);
      const expiresAt = Date.now() + tokens.expires_in * 1000;
      const scopeArray = tokens.scope.split(' ').filter(Boolean);

      this.persistTokens({
        accessToken: tokens.access_token,
        refreshToken: tokens.refresh_token,
        expiresAt,
        scope: scopeArray,
        tokenType: tokens.token_type,
        userId: profile.id,
        profile,
      });

      return {
        provider: 'twitter',
        result: {
          accessToken: {
            token: tokens.access_token,
            tokenType: tokens.token_type,
            expires: new Date(expiresAt).toISOString(),
            userId: profile.id,
          },
          refreshToken: tokens.refresh_token,
          scope: scopeArray,
          tokenType: tokens.token_type,
          expiresIn: tokens.expires_in,
          profile,
        },
      };
    } catch (err) {
      if (err instanceof Error) {
        return { error: err.message };
      }
      return { error: 'Twitter login failed unexpectedly.' };
    } finally {
      localStorage.removeItem(BaseSocialLogin.OAUTH_STATE_KEY);
    }
  }

  private async exchangeAuthorizationCode(code: string, pending: TwitterPendingLogin): Promise<TwitterTokenResponse> {
    const params = new URLSearchParams({
      grant_type: 'authorization_code',
      client_id: this.clientId ?? '',
      code,
      redirect_uri: pending.redirectUri,
      code_verifier: pending.codeVerifier,
    });

    const response = await fetch('https://api.x.com/2/oauth2/token', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: params.toString(),
    });

    if (!response.ok) {
      const text = await response.text();
      throw new Error(`Twitter token exchange failed (${response.status}): ${text}`);
    }

    return (await response.json()) as TwitterTokenResponse;
  }

  private async refreshWithRefreshToken(refreshToken: string): Promise<void> {
    const params = new URLSearchParams({
      grant_type: 'refresh_token',
      refresh_token: refreshToken,
      client_id: this.clientId ?? '',
    });

    const response = await fetch('https://api.x.com/2/oauth2/token', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: params.toString(),
    });

    if (!response.ok) {
      const text = await response.text();
      throw new Error(`Twitter refresh failed (${response.status}): ${text}`);
    }

    const tokens = (await response.json()) as TwitterTokenResponse;
    const profile = await this.fetchProfile(tokens.access_token);
    const expiresAt = Date.now() + tokens.expires_in * 1000;
    const scopeArray = tokens.scope.split(' ').filter(Boolean);

    this.persistTokens({
      accessToken: tokens.access_token,
      refreshToken: tokens.refresh_token ?? refreshToken,
      expiresAt,
      scope: scopeArray,
      tokenType: tokens.token_type,
      userId: profile.id,
      profile,
    });
  }

  private async fetchProfile(accessToken: string): Promise<TwitterProfile> {
    const fields = ['profile_image_url', 'verified', 'name', 'username'];
    const response = await fetch(`https://api.x.com/2/users/me?user.fields=${fields.join(',')}`, {
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
    });

    if (!response.ok) {
      const text = await response.text();
      throw new Error(`Unable to fetch Twitter profile (${response.status}): ${text}`);
    }

    const payload = (await response.json()) as TwitterUserResponse;
    if (!payload.data) {
      throw new Error('Twitter profile payload is missing data.');
    }

    return {
      id: payload.data.id,
      username: payload.data.username,
      name: payload.data.name ?? null,
      profileImageUrl: payload.data.profile_image_url ?? null,
      verified: payload.data.verified ?? false,
      email: (payload.data as { email?: string }).email ?? null,
    };
  }

  private persistTokens(tokens: TwitterStoredTokens): void {
    localStorage.setItem(this.TOKENS_KEY, JSON.stringify(tokens));
  }

  private getStoredTokens(): TwitterStoredTokens | null {
    const raw = localStorage.getItem(this.TOKENS_KEY);
    if (!raw) {
      return null;
    }
    try {
      return JSON.parse(raw) as TwitterStoredTokens;
    } catch (err) {
      console.warn('Failed to parse stored Twitter tokens', err);
      return null;
    }
  }

  private persistPendingLogin(state: string, payload: TwitterPendingLogin): void {
    localStorage.setItem(`${this.STATE_PREFIX}${state}`, JSON.stringify(payload));
  }

  private consumePendingLogin(state: string): TwitterPendingLogin | null {
    const key = `${this.STATE_PREFIX}${state}`;
    const raw = localStorage.getItem(key);
    localStorage.removeItem(key);
    if (!raw) {
      return null;
    }
    try {
      return JSON.parse(raw) as TwitterPendingLogin;
    } catch (err) {
      console.warn('Failed to parse pending Twitter login payload', err);
      return null;
    }
  }

  private generateState(): string {
    return [...crypto.getRandomValues(new Uint8Array(16))].map((b) => b.toString(16).padStart(2, '0')).join('');
  }

  private generateCodeVerifier(): string {
    const array = new Uint8Array(64);
    crypto.getRandomValues(array);
    return Array.from(array)
      .map((b) => 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-._~'[b % 66])
      .join('');
  }

  private async generateCodeChallenge(codeVerifier: string): Promise<string> {
    const encoder = new TextEncoder();
    const data = encoder.encode(codeVerifier);
    const digest = await crypto.subtle.digest('SHA-256', data);
    return this.base64UrlEncode(new Uint8Array(digest));
  }

  private base64UrlEncode(buffer: Uint8Array): string {
    let binary = '';
    buffer.forEach((b) => (binary += String.fromCharCode(b)));
    return btoa(binary).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
  }
}
